### ISSUE TITLE
Proof of Concept: Layered images

### ISSUE THREAD

**Issue by grembo** [pizzamig/pot#148]
# Layered Images

## Problem
I've been working on automatic staging using pot + nomad and I've been suffering from long turnaround times, rebuilding images, tarring them up, moving them to the registry and - finally - get them deployed. Between iterations, all that changed was the bespoken updated software installed in the image, so most of it (the OS, some base packages, etc.) stay the same. Lots of overhead, lots of disk space used (caches etc.). Wasteful.

## Idea
That's why I thought it would be nice, if images could be layered ("incremental"). Fortunately, it turns out that this is possible by combining  ZFS clones and ZFS snapshots, so I managed to successfully create a proof of concept that demonstrates how this works. I was actually able to integrate it into our experimental build pipeline.

## Concept
The basic concept is to change import/export to allow creating and importing images based on clones.

### Creation and Export
1. Create _single_ "foundation" image based on a base pot.
2. Install heavy dependencies in that foundation image.
3. Snapshot and export that foundation image.
4. Clone that image into a "leaf" image. 
5. Do some modifications of "leaf" image.
6. Export "leaf image" - detect that the image is a clone and include information about its origin in a meta file.
7. Include meta file information in skein signature file.

_Note: I rolled my custom implementation of applying flavors in this case, but thanks to recent changes, flavors can also be applied when using clones, so this works nicely._

In order to understand the _tag_ (=version) of the last export of the dependency pot, ZFS properties are set on the file system backing the origin.

### Import
On import image, do:
1. Check if image depends on an origin (found in meta file).
2. Check if that dependency is already installed. If not, install it from the same base URL.
3. Load leaf image snapshot, referencing the correct origin.

## The Code
Code changes [can be found here](https://github.com/grembo/pot/compare/4527f5c..layered-images). The implementation is limited in that it only supports straightforward pot setups (thats is, only two ZFS filesystems, e.g., _/opt/pot/jails/name_ and _/opt/pot/jails/name/m_. Only tested on FreeBSD 12. I didn't open a pull request right away, as I would like to get your opinion on this first.

### Very basic example
(assumes nginx on localhost)

#### Create/Export
```sh
pot create -p foundation -t single -b 12.2
echo "Inside foundation" >/opt/pot/jails/foundation/m/whereami
pot snapshot -p foundation

pot clone -P foundation -p leaf
echo "Inside leaf" >/opt/pot/jails/leaf/m/whereami
pot snapshot -p leaf

pot export -D /usr/local/www/nginx -p foundation -t 2021Q2 -A
pot export -D /usr/local/www/nginx -p leaf -t 20210516 -A
```

#### Import
This is tested using `pot prepare`, as this is what the nomad plugin is using as well.

```sh
pot prepare -U http://localhost -p leaf -t 20210516 -a myalloc
```

Output:
```
===>  importing leaf @ 20210516 as leaf_20210516
/var/cache/pot/leaf_20210516.xz                       7128  B   93 MBps    00s
/var/cache/pot/leaf_20210516.xz.skein                  257  B 3927 kBps    00s
/var/cache/pot/leaf_20210516.xz.meta                    29  B  471 kBps    00s
===>  Pot leaf_20210516 depends on foundation_2021Q2 (@1621123854)
===>  Installing dependency foundation_2021Q2
===>  importing foundation @ 2021Q2 as foundation_2021Q2
/var/cache/pot/foundation_2021Q2.xz                    179 MB   71 MBps    02s
/var/cache/pot/foundation_2021Q2.xz.skein              257  B 3987 kBps    00s
/var/cache/pot/foundation_2021Q2.xz.meta                 2  B   33 kBps    00s
===>  Pot foundation_2021Q2 has no dependencies
===>  Prepared the pot leaf_20210516_myalloc
```

Check files:
```sh
$ cat /opt/pot/jails/foundation_2021Q2/m/whereami
Inside foundation
$ cat /opt/pot/jails/leaf_20210516_myalloc/m/whereami 
Inside leaf
```

#### Updating
```sh
echo "Inside leaf updated" >/opt/pot/jails/leaf/m/whereami
pot snapshot -p leaf
pot export -D /usr/local/www/nginx -p leaf -t 20210516_2 -A

pot prepare -U http://localhost -p leaf -t 20210516_2 -a myalloc
```

Output:
```
pot prepare -U http://localhost -p leaf -t 20210516_2 -a myalloc
===>  importing leaf @ 20210516_2 as leaf_20210516_2
/var/cache/pot/leaf_20210516_2.xz                     4964  B   65 MBps    00s
/var/cache/pot/leaf_20210516_2.xz.skein                257  B 3927 kBps    00s
/var/cache/pot/leaf_20210516_2.xz.meta                  29  B  473 kBps    00s
===>  Pot leaf_20210516_2 depends on foundation_2021Q2 (@1621124565)
===>  foundation_2021Q2 already installed
===>  Prepared the pot leaf_20210516_2_myalloc
```

### Using Flavors
Using flavo(u)rs is easy, thanks to recent changes in `pot clone`:

```sh
# pot clone -P foundation -p yummy -f fbsd-update
===>  Starting the pot yummy
ELF ldconfig path: /lib /usr/lib /usr/lib/compat
32-bit compatibility ldconfig path: /usr/lib32
Generating host.conf.
Creating and/or trimming log files.
Starting syslogd.
Clearing /tmp (X related).
Updating motd:.
Updating /var/run/os-release done.
Starting sendmail_submit.
Starting sendmail_msp_queue.
Starting cron.

Sun May 16 00:32:36 UTC 2021
/usr/local/etc/pot/flavours/fbsd-update.sh -> /opt/pot/jails/yummy/m/tmp/fbsd-update.sh
src component not installed, skipped
Looking up update.FreeBSD.org mirrors... 2 mirrors found.
Fetching public key from update1.freebsd.org... done.
Fetching metadata signature for 12.2-RELEASE from update1.freebsd.org... done.
Fetching metadata index... done.
Fetching 2 metadata files... done.
Inspecting system... done.
Preparing to download files... done.
Fetching 72 patches.....10....20....30....40....50....60....70. done.
Applying patches... done.
Fetching 1 files...  done.
The following files will be removed as part of updating to
12.2-RELEASE-p6:
......
```

### Funny Example
This example shows that this can be taken to extremes:
```sh
pot create -p foundation -t single -b 12.2
pot snapshot -p foundation

pot clone -P foundation -p level1
pot snapshot -p level1
pot clone -P level1 -p level2
pot snapshot -p level2
pot clone -P level2 -p level3
pot snapshot -p level3

pot export -D /usr/local/www/nginx -p foundation -t 2021Q2
pot export -D /usr/local/www/nginx -p level1 -t 20210516
pot export -D /usr/local/www/nginx -p level2 -t 20210516
pot export -D /usr/local/www/nginx -p level3 -t 20210516

pot import -U http://localhost -p level3 -t 20210516
```

Import Output:
```
===>  importing level3 @ 20210516 as level3_20210516
/var/cache/pot/level3_20210516.xz                     3908  B   52 MBps    00s
/var/cache/pot/level3_20210516.xz.skein                257  B 4112 kBps    00s
/var/cache/pot/level3_20210516.xz.meta                  27  B  438 kBps    00s
===>  Pot level3_20210516 depends on level2_20210516 (@1621127437)
===>  Installing dependency level2_20210516
===>  importing level2 @ 20210516 as level2_20210516
/var/cache/pot/level2_20210516.xz                     3904  B   53 MBps    00s
/var/cache/pot/level2_20210516.xz.skein                257  B 4128 kBps    00s
/var/cache/pot/level2_20210516.xz.meta                  27  B  445 kBps    00s
===>  Pot level2_20210516 depends on level1_20210516 (@1621127436)
===>  Installing dependency level1_20210516
===>  importing level1 @ 20210516 as level1_20210516
/var/cache/pot/level1_20210516.xz                     3900  B   54 MBps    00s
/var/cache/pot/level1_20210516.xz.skein                257  B 4112 kBps    00s
/var/cache/pot/level1_20210516.xz.meta                  29  B  471 kBps    00s
===>  Pot level1_20210516 depends on foundation_2021Q2 (@1621127434)
===>  Installing dependency foundation_2021Q2
===>  importing foundation @ 2021Q2 as foundation_2021Q2
/var/cache/pot/foundation_2021Q2.xz                    178 MB   58 MBps    03s
/var/cache/pot/foundation_2021Q2.xz.skein              257  B 3987 kBps    00s
/var/cache/pot/foundation_2021Q2.xz.meta                 2  B   32 kBps    00s
===>  Pot foundation_2021Q2 has no dependencies
```

## Conclusion
The patches aren't perfect for sure (probably contain bugs) and one has to be a bit careful to not mess up the versions of the different layers and make sure things are destructed in the correct order.

That said, this works in principle and having this feature is really useful, especially when iterating quickly over images in a staging environment. Using a foundation image (plus potentially one intermediate one) that has all the heavy dependencies, rolling a new staging image from development in our environment was sped up by >80%, saving a lot of space in the process. These benefits multiply, once one has many images based on the same foundation running on one jailhost/nomad compute node.

**Comment by grembo** [comment]
To make working with this easier , I added a new flag to `pot export`:

```
  -c : Treat tags as versions and check if they are decreasing
```

When specified, this does:
1. Check if the tag (sorted as a version number) is >= the one of the last export
2. If the tag == the one of the last export, check if it refers to the same snapshot

**Comment by pizzamig** [comment]
Hi.
Thanks for working on this.
The flavor support to clone was exactly intended for this use case, even if the current support is limited to local pots and not to images.
The support for export/import is the problem I didn't have time to face and it seems you are solving adding a meta file to carry layer information, a good way to solve it, IMO.

Over the weekend, I'll prepare a new release of pot (there are a lot of changes not yet released) and then I'll be more than happy to merge your work in main and refine the implementations in master. Is it OK for you?

**Comment by grembo** [comment]
> Hi.
> Thanks for working on this.
> The flavor support to clone was exactly intended for this use case, even if the current support is limited to local pots and not to images.
> The support for export/import is the problem I didn't have time to face and it seems you are solving adding a meta file to carry layer information, a good way to solve it, IMO.
> 
> Over the weekend, I'll prepare a new release of pot (there are a lot of changes not yet released) and then I'll be more than happy to merge your work in main and refine the implementations in master. Is it OK for you?

A new release sounds like a good idea, lots of changes to pot in the last couple of months ^_^

Feel free to get in touch with me to discuss details about my proof of concept - I can also give you some more context about my intended use case, which in turn might give you more ideas for improvements.

**Comment by grembo** [comment]
@pizzamig I think this can be closed now (we landed initial support of the feature in #151).

**Comment by pizzamig** [comment]
I agree.
Closing it
